### PR TITLE
Vendor jemalloc and use it for CPU Tensor allocations

### PR DESCRIPTION
**Pull request by peterbell10**
Resolves https://github.com/pytorch/pytorch/pull/53715#issuecomment-815077847

This links a static version of jemalloc and uses it for C10's default cpu allocator. Note that jemalloc currently uses autoconf for its build system, though I saw some talk of a future move to CMake in their github issues.

### CODE REVIEW

**Comment by ezyang** [comment]
It'll be good to talk about how we plan to benchmark this change.

**Comment by peterbell10** [comment]
I've been running a few ad-hoc timeit benchmarks with `torch.empty`:
```python
import torch

%timeit torch.empty(10)
%timeit torch.empty(100)
%timeit torch.empty(1000)
```

I observed that using `jemalloc` was around 5-10% faster overall, so even after including python overhead.  I also saw that my system's `posix_memalign` was even slower for some sizes (100 in this case), whereas `jemalloc` seemed to be completely consistent across sizes.

Allocating the same size in a loop isn't a very realistic scenario though, so something doing more diverse allocations would be good. I do worry something like a real NN would have too much variance in runtime to reliably compare though.

Thoughts @ezyang?

**Comment by ezyang** [comment]
Maybe @colesbury has some more thoughts but I think the most realistic thing to do is benchmark a few important CPU models, esp language models. You can find some in https://github.com/pytorch/benchmark

**Comment by ezyang** [comment]
@swolchok can you confirm that this change isn't relevant to FB internal, since we probably use a better malloc by default anyway?

**Comment by colesbury** [comment]
For synthetic benchmarks, I'd like to see timings for both `torch.empty` AND `torch.zeros` (or `torch.ones`) over a range of sizes up to ~100 MB. (Powers of 10, like you were doing, seem reasonable). I would expect a substantial improvement for jemalloc over glibc for `torch.zeros`/`torch.ones` on sizes in the megabytes or tens of megabytes. From what I remember, glibc malloc tends to be more aggressive with `mmap`/`munmap`; that cost usually shows up in the first access (so in `torch.zeros` and `torch.ones`, but not `torch.empty`).

For more representative benchmarks, @ezyang's suggestion of CPU language models makes sense.

**Comment by swolchok** [comment]
> @swolchok can you confirm that this change isn't relevant to FB internal, since we probably use a better malloc by default anyway?

not particularly familiar with this, but I think that's more or less accurate

**Comment by ezyang** [comment]
Any news on this? :)

**Comment by peterbell10** [comment]
Oh right, when I ran the benchmarks evaluation mode showed an improvement but training was consistently worse. IIRC ~10-20% slow down.

**Comment by ezyang** [comment]
RIP
